### ISSUE TITLE
[Feature request] Integration with Bilibili/flv.js for FLV playback and livestream

### ISSUE THREAD

**Issue by m13253** [flowplayer/flowplayer#1074]
https://github.com/bilibili/flv.js is a new JavaScript library to play and stream FLV videos.

It supports:
- FLV container with H.264 + AAC codec playback
- Multipart segmented video playback
- HTTP FLV low latency live stream playback
- FLV over WebSocket live stream playback
- Compatible with Chrome, FireFox, Safari 10, IE11 and Edge
- Extermely low overhead, and hardware accelerated by your browser

I hope flv.js can integrate with Flowplayer to provide native FLV playback without the support of Flash.

Also, HTTP FLV live stream provides much lower latency than HLS or DASH, and is chosen by many leading livestream sites (at least in China, e.g. Bilibili.com, Douyu.tv, Panda.tv). By integrating with flv.js can we provide lower latency stream with pure HTML5.

I am ready to help with the integration.

**Comment by blacktrash** [comment]
I'm curious:
Why is it that this approach has less latency? I am aware that RTMP features lower latency, but I wonder - lack of experience in live deployment - whether this is caused on the _server_ side because servers traditionally mux to FLV first? Why is there a delay in muxing e.g. to HLS? Can't be the required advance buffer ensuring smooth playback, that problem would arise with other containers too, or wouldn't it?

**Comment by m13253** [comment]
HTTP FLV has the same latency with RTMP. One difference is RTMP has latency control (the ability to know the latency) while HTTP FLV does not. But FLV over WebSocket might have latency control.

HTTP FLV, alongside with RTMP, is supported by many Chinese CDNs (ChinaCache, ChinaNetCenter, etc) and some [open source RTMP server](https://github.com/ossrs/srs/wiki/v2_EN_DeliveryHttpStream#about-http-flv).
I live in China so I don't know the worldwide usage, but HTTP FLV is indeed a good choice and requires less complicated CDN configuration (than RTMP).

HLS splits stream into `.ts` files, while `hls.js` picks up the segments to remux them into `.mp4` blobs. The problem of MP4 is the muxer can't write `moov` atom before everything else is finished; the player can't play the stream before `moov` atom is read: this will introduce a latency of `2 * segmentLength`.
Also `hls.js` reads a full segment before it can do remux, this will introduce another latency of `1 * segmentLength`.
From my test on a Local Area Network, an H.264+AAC stream with GOP = 2 sec, HLS Fragment = 10 sec, will cause `hls.js` to have a latency of 30 sec - 45 sec. While HTTP FLV has a latency of less than 15 sec or lower.

**Comment by m13253** [comment]
The author of SRS has written this:

**Why HTTP FLV**
- The internet realtime live streaming, generally use RTMP for low latence, while HTTP FLV latency equals to RTMP.
- Firewall traverse: Many firewall may block RTMP, but never HTTP.
- For GSLB: The HTTP FLV can support HTTP 302, to help GSLB to correct the stupid DNS problem.
- Fault tolerance: The HTTP FLV of SRS similar to RTMP, support mulitple origin and fault backup.
- Common Protocol: The HTTP FLV is ok for flash and other player.
- Simple: FLV is the most simple media stream fomat, HTTP is the standard protocol of internet, they make thinks simple.

I think HTTP FLV is a good point for flv.js.
Another good thing is we have one less reason to use fallback Flash player to play videos "just because the video was served in FLV format".

**Comment by blacktrash** [comment]
That doesn't really answer my question. DASH and HLS also HTTP, and your answer doesn't explain the delay. HLS offers redundant stream failover etc. As for simplicity of the container format: it can't be simpler than MPEG2-TS which HLS uses - which is also why it still has some trouble in MacOS Safari ironically - DASH is more advanced uses a - more complex - MP4 container which is in fact easier to process for MSE, whereas hls.js has to remux the simple format for MediaSource.
Anyway, I'm still curious regarding the delay - the above doesn't answer that, does it?

Other than that it should be relatively easy to bake this into an engine plugin like hlsjs or dashjs. The problems are more of trivial/practical nature:
- maintenance time combined with
- live streams for testing hard to come by in a reliable fashion
- similar to dash this would require additional streams to be platform/browser compatible, even when ignoring IE8 and Win7 where MediaSource is not supported, there's iOS ... that's why from our perspective the dashjs plugin is a bit on the back burner (combined with lack of time).

In other words: you've applied to be the maintainer of a flvjs plugin ;-) Should be pretty similar to [hlsjs](https://github.com/flowplayer/flowplayer-hlsjs) and/or [dashjs](https://github.com/flowplayer/flowplayer-mpegdash), so you can probably take those as a starting point.
If/when it's reasonable stable we'd be happy to advertise it at https://flowplayer.org/docs/plugins.html

**Comment by m13253** [comment]
> Anyway, I'm still curious regarding the delay - the above doesn't answer that, does it?

Let me repost my words:
HLS splits stream into .ts files, while hls.js picks up the segments to remux them into .mp4 blobs. The problem of MP4 is the muxer can't write moov atom before everything else is finished; the player can't play the stream before moov atom is read: this will introduce a latency of `2 * segmentLength`. Also hls.js reads a full segment before it can do remux, this will add another latency of `1 * segmentLength`.

That's why hls.js or dash.js have a latency of at least `3 * segmentLength + 2 * GOP`.

@xqq (the author of flv.js) just pointed out that flv.js can reach a latency of less than 1 second. It is because no segmentation is happened -- the replay happens immediately. Neither the remuxer or the player does not need to wait for a "fragmentation", and can start all the work as soon as the first byte is received.

**Comment by blacktrash** [comment]
ok, that makes sense, also why there would still be larger delay in generic HLS playback.

So, an engine plugin would be welcome ;-)

Just to be certain, you should double check whether it works out of the box by pretending source type "video/mp4" and pretending to the engine that the stream is not live, but disguising it as live UI-wise via CSS done for the icecast audio stream here: http://demos.flowplayer.org/api/audio.html#icecast

**Comment by nnarhinen** [comment]
Whoever wants to implement a plugin for flv.js, feel free too do that!

When it's in a state where it is generally usable for a wider audience, drop a note in support@flowplayer.org and we'll review it and add to our plugins page.

Closing the issue here, as it will not be part of core.
